In [2]:
import pandas as pd

In [3]:
df1 = pd.read_excel('/Users/constanzavargasmier/Desktop/Ale/DATASETS/New Customers.xlsx', sheet_name = 'January')
df1

ID  Joining Day    Demographic      Value
0    490910            3      Ethnicity      White
1    490910            3  Date of Birth  5/23/1981
2    490910            3   Account Type      Basic
3    369221           18      Ethnicity      Black
4    369221           18  Date of Birth   3/4/2019
..      ...          ...            ...        ...
244  840464            6  Date of Birth  7/18/1968
245  840464            6   Account Type       Gold
246  674869           13      Ethnicity      Asian
247  674869           13  Date of Birth   6/6/1991
248  674869           13   Account Type   Platinum

[249 rows x 4 columns]

In [4]:
df2 = pd.read_excel('/Users/constanzavargasmier/Desktop/Ale/DATASETS/New Customers.xlsx', sheet_name = 'February')
df3 = pd.read_excel('/Users/constanzavargasmier/Desktop/Ale/DATASETS/New Customers.xlsx', sheet_name = 'March')
df4 = pd.read_excel('/Users/constanzavargasmier/Desktop/Ale/DATASETS/New Customers.xlsx', sheet_name = 'April')
df5 = pd.read_excel('/Users/constanzavargasmier/Desktop/Ale/DATASETS/New Customers.xlsx', sheet_name = 'May')
df6 = pd.read_excel('/Users/constanzavargasmier/Desktop/Ale/DATASETS/New Customers.xlsx', sheet_name = 'June')
df7 = pd.read_excel('/Users/constanzavargasmier/Desktop/Ale/DATASETS/New Customers.xlsx', sheet_name = 'July')
df8 = pd.read_excel('/Users/constanzavargasmier/Desktop/Ale/DATASETS/New Customers.xlsx', sheet_name = 'August')
df9 = pd.read_excel('/Users/constanzavargasmier/Desktop/Ale/DATASETS/New Customers.xlsx', sheet_name = 'September')
df10 = pd.read_excel('/Users/constanzavargasmier/Desktop/Ale/DATASETS/New Customers.xlsx', sheet_name = 'October')
df11 = pd.read_excel('/Users/constanzavargasmier/Desktop/Ale/DATASETS/New Customers.xlsx', sheet_name = 'November')
df12 = pd.read_excel('/Users/constanzavargasmier/Desktop/Ale/DATASETS/New Customers.xlsx', sheet_name = 'December')

In [15]:
df12

ID  Joining Day    Demographic       Value
0    931463           22      Ethnicity       Other
1    931463           22  Date of Birth   8/30/1978
2    931463           22   Account Type    Platinum
3    973596           16      Ethnicity       Other
4    973596           16  Date of Birth   2/22/1971
..      ...          ...            ...         ...
268  174699            2  Date of Birth   3/13/1989
269  174699            2   Account Type        Gold
270  514598           28      Ethnicity       Other
271  514598           28  Date of Birth  10/10/1971
272  514598           28   Account Type    Platinum

[273 rows x 4 columns]

### Here I ran df for all the dataframes, to check if they were using the same columns

In [19]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 249 entries, 0 to 248
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   ID           249 non-null    int64 
 1   Joining Day  249 non-null    int64 
 2   Demographic  249 non-null    object
 3   Value        249 non-null    object
dtypes: int64(2), object(2)
memory usage: 7.9+ KB


In [20]:
df1['Month'] = 1

In [21]:
df1

ID  Joining Day    Demographic      Value  Month
0    490910            3      Ethnicity      White      1
1    490910            3  Date of Birth  5/23/1981      1
2    490910            3   Account Type      Basic      1
3    369221           18      Ethnicity      Black      1
4    369221           18  Date of Birth   3/4/2019      1
..      ...          ...            ...        ...    ...
244  840464            6  Date of Birth  7/18/1968      1
245  840464            6   Account Type       Gold      1
246  674869           13      Ethnicity      Asian      1
247  674869           13  Date of Birth   6/6/1991      1
248  674869           13   Account Type   Platinum      1

[249 rows x 5 columns]

In [22]:
df2['Month'] = 2
df3['Month'] = 3
df4['Month'] = 4
df5['Month'] = 5
df6['Month'] = 6
df7['Month'] = 7
df8['Month'] = 8
df9['Month'] = 9
df10['Month'] = 10
df11['Month'] = 11
df12['Month'] = 12


In [23]:
df1

ID  Joining Day    Demographic      Value  Month
0    490910            3      Ethnicity      White      1
1    490910            3  Date of Birth  5/23/1981      1
2    490910            3   Account Type      Basic      1
3    369221           18      Ethnicity      Black      1
4    369221           18  Date of Birth   3/4/2019      1
..      ...          ...            ...        ...    ...
244  840464            6  Date of Birth  7/18/1968      1
245  840464            6   Account Type       Gold      1
246  674869           13      Ethnicity      Asian      1
247  674869           13  Date of Birth   6/6/1991      1
248  674869           13   Account Type   Platinum      1

[249 rows x 5 columns]

In [27]:
df1['Year'] = 23
df2['Year'] = 23
df3['Year'] = 23
df4['Year'] = 23
df5['Year'] = 23
df6['Year'] = 23
df7['Year'] = 23
df8['Year'] = 23
df9['Year'] = 23
df10['Year'] = 23
df11['Year'] = 23
df12['Year'] = 23

In [29]:
df1

ID  Joining Day    Demographic      Value  Month  Year
0    490910            3      Ethnicity      White      1    23
1    490910            3  Date of Birth  5/23/1981      1    23
2    490910            3   Account Type      Basic      1    23
3    369221           18      Ethnicity      Black      1    23
4    369221           18  Date of Birth   3/4/2019      1    23
..      ...          ...            ...        ...    ...   ...
244  840464            6  Date of Birth  7/18/1968      1    23
245  840464            6   Account Type       Gold      1    23
246  674869           13      Ethnicity      Asian      1    23
247  674869           13  Date of Birth   6/6/1991      1    23
248  674869           13   Account Type   Platinum      1    23

[249 rows x 6 columns]

In [30]:
df1['Joining_Date'] = df1['Joining Day'].astype(str) + '/' + df1['Month'].astype(str) + '/' + df1['Year'].astype(str)

In [31]:
df1

ID  Joining Day    Demographic      Value  Month  Year Joining_Date
0    490910            3      Ethnicity      White      1    23       3/1/23
1    490910            3  Date of Birth  5/23/1981      1    23       3/1/23
2    490910            3   Account Type      Basic      1    23       3/1/23
3    369221           18      Ethnicity      Black      1    23      18/1/23
4    369221           18  Date of Birth   3/4/2019      1    23      18/1/23
..      ...          ...            ...        ...    ...   ...          ...
244  840464            6  Date of Birth  7/18/1968      1    23       6/1/23
245  840464            6   Account Type       Gold      1    23       6/1/23
246  674869           13      Ethnicity      Asian      1    23      13/1/23
247  674869           13  Date of Birth   6/6/1991      1    23      13/1/23
248  674869           13   Account Type   Platinum      1    23      13/1/23

[249 rows x 7 columns]

In [32]:
df2['Joining_Date'] = df2['Joining Day'].astype(str) + '/' + df2['Month'].astype(str) + '/' + df2['Year'].astype(str)
df3['Joining_Date'] = df3['Joining Day'].astype(str) + '/' + df3['Month'].astype(str) + '/' + df3['Year'].astype(str)
df4['Joining_Date'] = df4['Joining Day'].astype(str) + '/' + df4['Month'].astype(str) + '/' + df4['Year'].astype(str)
df5['Joining_Date'] = df5['Joining Day'].astype(str) + '/' + df5['Month'].astype(str) + '/' + df5['Year'].astype(str)
df6['Joining_Date'] = df6['Joining Day'].astype(str) + '/' + df6['Month'].astype(str) + '/' + df6['Year'].astype(str)
df7['Joining_Date'] = df7['Joining Day'].astype(str) + '/' + df7['Month'].astype(str) + '/' + df7['Year'].astype(str)
df8['Joining_Date'] = df8['Joining Day'].astype(str) + '/' + df8['Month'].astype(str) + '/' + df8['Year'].astype(str)
df9['Joining_Date'] = df9['Joining Day'].astype(str) + '/' + df9['Month'].astype(str) + '/' + df9['Year'].astype(str)
df10['Joining_Date'] = df10['Joining Day'].astype(str) + '/' + df10['Month'].astype(str) + '/' + df10['Year'].astype(str)
df11['Joining_Date'] = df11['Joining Day'].astype(str) + '/' + df11['Month'].astype(str) + '/' + df11['Year'].astype(str)
df12['Joining_Date'] = df12['Joining Day'].astype(str) + '/' + df12['Month'].astype(str) + '/' + df12['Year'].astype(str)

In [33]:
df1

ID  Joining Day    Demographic      Value  Month  Year Joining_Date
0    490910            3      Ethnicity      White      1    23       3/1/23
1    490910            3  Date of Birth  5/23/1981      1    23       3/1/23
2    490910            3   Account Type      Basic      1    23       3/1/23
3    369221           18      Ethnicity      Black      1    23      18/1/23
4    369221           18  Date of Birth   3/4/2019      1    23      18/1/23
..      ...          ...            ...        ...    ...   ...          ...
244  840464            6  Date of Birth  7/18/1968      1    23       6/1/23
245  840464            6   Account Type       Gold      1    23       6/1/23
246  674869           13      Ethnicity      Asian      1    23      13/1/23
247  674869           13  Date of Birth   6/6/1991      1    23      13/1/23
248  674869           13   Account Type   Platinum      1    23      13/1/23

[249 rows x 7 columns]

In [42]:
df13 = pd.concat([df1, df2, df3, df4, df5, df6, df7, df8, df9, df10, df11, df12], axis=0)
df13

ID  Joining Day    Demographic       Value  Month  Year Joining_Date  \
0    490910            3      Ethnicity       White      1    23       3/1/23   
1    490910            3  Date of Birth   5/23/1981      1    23       3/1/23   
2    490910            3   Account Type       Basic      1    23       3/1/23   
3    369221           18      Ethnicity       Black      1    23      18/1/23   
4    369221           18  Date of Birth    3/4/2019      1    23      18/1/23   
..      ...          ...            ...         ...    ...   ...          ...   
268  174699            2  Date of Birth   3/13/1989     12    23      2/12/23   
269  174699            2   Account Type        Gold     12    23      2/12/23   
270  514598           28      Ethnicity       Other     12    23     28/12/23   
271  514598           28  Date of Birth  10/10/1971     12    23     28/12/23   
272  514598           28   Account Type    Platinum     12    23     28/12/23   

    Demographiic Demagraphic  
0            NaN         NaN  
1            NaN         NaN  
2            NaN         NaN  
3            NaN         NaN  
4            NaN         NaN  
..           ...         ...  
268          NaN         NaN  
269          NaN         NaN  
270          NaN         NaN  
271          NaN         NaN  
272          NaN         NaN  

[2970 rows x 9 columns]

### combining the df I realized some of the demographic columns are not written correctly

In [53]:
df8.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 258 entries, 0 to 257
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   ID            258 non-null    int64 
 1   Joining Day   258 non-null    int64 
 2   Demographiic  258 non-null    object
 3   Value         258 non-null    object
 4   Month         258 non-null    int64 
 5   Year          258 non-null    int64 
 6   Joining_Date  258 non-null    object
dtypes: int64(4), object(3)
memory usage: 14.2+ KB


In [55]:
df10.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 276 entries, 0 to 275
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   ID            276 non-null    int64 
 1   Joining Day   276 non-null    int64 
 2   Demagraphic   276 non-null    object
 3   Value         276 non-null    object
 4   Month         276 non-null    int64 
 5   Year          276 non-null    int64 
 6   Joining_Date  276 non-null    object
dtypes: int64(4), object(3)
memory usage: 15.2+ KB


In [58]:
df8.rename(columns={'Demographiic': 'Demographic'}, inplace=True)

In [60]:
df10.rename(columns={'Demagraphic': 'Demographic'}, inplace=True)

In [62]:
df14 = pd.concat([df1, df2, df3, df4, df5, df6, df7, df8, df9, df10, df11, df12], axis=0)
df14

ID  Joining Day    Demographic       Value  Month  Year Joining_Date
0    490910            3      Ethnicity       White      1    23       3/1/23
1    490910            3  Date of Birth   5/23/1981      1    23       3/1/23
2    490910            3   Account Type       Basic      1    23       3/1/23
3    369221           18      Ethnicity       Black      1    23      18/1/23
4    369221           18  Date of Birth    3/4/2019      1    23      18/1/23
..      ...          ...            ...         ...    ...   ...          ...
268  174699            2  Date of Birth   3/13/1989     12    23      2/12/23
269  174699            2   Account Type        Gold     12    23      2/12/23
270  514598           28      Ethnicity       Other     12    23     28/12/23
271  514598           28  Date of Birth  10/10/1971     12    23     28/12/23
272  514598           28   Account Type    Platinum     12    23     28/12/23

[2970 rows x 7 columns]

In [64]:
df14 = df14.drop(columns = ['Joining Day', 'Month', 'Year'])
df14

ID    Demographic       Value Joining_Date
0    490910      Ethnicity       White       3/1/23
1    490910  Date of Birth   5/23/1981       3/1/23
2    490910   Account Type       Basic       3/1/23
3    369221      Ethnicity       Black      18/1/23
4    369221  Date of Birth    3/4/2019      18/1/23
..      ...            ...         ...          ...
268  174699  Date of Birth   3/13/1989      2/12/23
269  174699   Account Type        Gold      2/12/23
270  514598      Ethnicity       Other     28/12/23
271  514598  Date of Birth  10/10/1971     28/12/23
272  514598   Account Type    Platinum     28/12/23

[2970 rows x 4 columns]

## Now I will filter Demographic to get every value and create new columns with those values

In [67]:
df15 = df14[df14['Demographic'] == 'Ethnicity']

In [68]:
df15

ID Demographic  Value Joining_Date
0    490910   Ethnicity  White       3/1/23
3    369221   Ethnicity  Black      18/1/23
6    568991   Ethnicity  White      16/1/23
9    191607   Ethnicity  White       9/1/23
12   612794   Ethnicity  Asian      29/1/23
..      ...         ...    ...          ...
258  970548   Ethnicity  Asian     12/12/23
261  569923   Ethnicity  Other      4/12/23
264  354147   Ethnicity  Black     20/12/23
267  174699   Ethnicity  Asian      2/12/23
270  514598   Ethnicity  Other     28/12/23

[990 rows x 4 columns]

In [70]:
df15 = df15.drop(columns = 'Demographic')

In [71]:
df15.rename(columns={'Value': 'Ethnicity'}, inplace=True)

In [72]:
df15

ID Ethnicity Joining_Date
0    490910     White       3/1/23
3    369221     Black      18/1/23
6    568991     White      16/1/23
9    191607     White       9/1/23
12   612794     Asian      29/1/23
..      ...       ...          ...
258  970548     Asian     12/12/23
261  569923     Other      4/12/23
264  354147     Black     20/12/23
267  174699     Asian      2/12/23
270  514598     Other     28/12/23

[990 rows x 3 columns]

In [74]:
df16 = df14[df14['Demographic'] == 'Date of Birth']

In [76]:
df16 = df16.drop(columns = ['Demographic','Joining_Date'])

In [78]:
df16 = df16.rename(columns = {'Value':'Date of Birth'})

In [79]:
df16

ID Date of Birth
1    490910     5/23/1981
4    369221      3/4/2019
7    568991     4/27/1979
10   191607     3/15/1963
13   612794    11/12/1981
..      ...           ...
259  970548      5/9/1996
262  569923    12/30/1975
265  354147     4/14/2019
268  174699     3/13/1989
271  514598    10/10/1971

[990 rows x 2 columns]

In [81]:
df17 = df14[df14['Demographic'] == 'Account Type']

In [83]:
df17 = df17.drop(columns = ['Demographic', 'Joining_Date'])

In [85]:
df17 = df17.rename(columns = {'Value':'Account Type'})

In [86]:
df17

ID Account Type
2    490910        Basic
5    369221     Platinum
8    568991         Gold
11   191607         Gold
14   612794        Basic
..      ...          ...
260  970548     Platinum
263  569923         Gold
266  354147         Gold
269  174699         Gold
272  514598     Platinum

[990 rows x 2 columns]

In [87]:
df15

ID Ethnicity Joining_Date
0    490910     White       3/1/23
3    369221     Black      18/1/23
6    568991     White      16/1/23
9    191607     White       9/1/23
12   612794     Asian      29/1/23
..      ...       ...          ...
258  970548     Asian     12/12/23
261  569923     Other      4/12/23
264  354147     Black     20/12/23
267  174699     Asian      2/12/23
270  514598     Other     28/12/23

[990 rows x 3 columns]

In [90]:
df18 = df15.set_index('ID').join(df16.set_index('ID'), lsuffix='_Left', how='inner').join(df17.set_index('ID'), lsuffix='_Left', how='inner')

In [91]:
df18

Ethnicity Joining_Date Date of Birth Account Type
ID                                                      
100185     Asian      20/5/23     7/29/1952        Basic
101515     Black      14/4/23     8/11/1974         Gold
101744     Asian      29/8/23     1/21/1945        Basic
102704     Black      23/1/23      3/9/2000        Basic
103488     Other      28/8/23     9/26/1957        Basic
...          ...          ...           ...          ...
994016     Other      29/3/23     3/29/1955     Platinum
994289     White      16/6/23      5/9/1990         Gold
994611     Black      10/1/23     6/19/1994        Basic
995456     Other      17/2/23      3/5/1975        Basic
997703     Other      19/9/23      1/7/1998     Platinum

[996 rows x 4 columns]

In [93]:
df19 = df18.reset_index()

In [94]:
df19.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 996 entries, 0 to 995
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   ID             996 non-null    int64 
 1   Ethnicity      996 non-null    object
 2   Joining_Date   996 non-null    object
 3   Date of Birth  996 non-null    object
 4   Account Type   996 non-null    object
dtypes: int64(1), object(4)
memory usage: 39.0+ KB


In [96]:
df19.nunique()

ID               989
Ethnicity          4
Joining_Date     340
Date of Birth    966
Account Type       3
dtype: int64

### With nunique I found there are duplicate ID because there are 996 entries on the column and 989 ID's

In [98]:
df19 = df19.drop_duplicates()

In [99]:
df19

ID Ethnicity Joining_Date Date of Birth Account Type
0    100185     Asian      20/5/23     7/29/1952        Basic
1    101515     Black      14/4/23     8/11/1974         Gold
2    101744     Asian      29/8/23     1/21/1945        Basic
3    102704     Black      23/1/23      3/9/2000        Basic
4    103488     Other      28/8/23     9/26/1957        Basic
..      ...       ...          ...           ...          ...
991  994016     Other      29/3/23     3/29/1955     Platinum
992  994289     White      16/6/23      5/9/1990         Gold
993  994611     Black      10/1/23     6/19/1994        Basic
994  995456     Other      17/2/23      3/5/1975        Basic
995  997703     Other      19/9/23      1/7/1998     Platinum

[990 rows x 5 columns]

In [101]:
df19 = df19.reset_index(drop=True)

In [102]:
df19

ID Ethnicity Joining_Date Date of Birth Account Type
0    100185     Asian      20/5/23     7/29/1952        Basic
1    101515     Black      14/4/23     8/11/1974         Gold
2    101744     Asian      29/8/23     1/21/1945        Basic
3    102704     Black      23/1/23      3/9/2000        Basic
4    103488     Other      28/8/23     9/26/1957        Basic
..      ...       ...          ...           ...          ...
985  994016     Other      29/3/23     3/29/1955     Platinum
986  994289     White      16/6/23      5/9/1990         Gold
987  994611     Black      10/1/23     6/19/1994        Basic
988  995456     Other      17/2/23      3/5/1975        Basic
989  997703     Other      19/9/23      1/7/1998     Platinum

[990 rows x 5 columns]